File: AudioKit/Operations/Signal Input and Output/AKLog.m
//
//  AKLog.m
//  AudioKit
//
//  Created by Aurelius Prochazka on 2/26/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "AKLog.h"

@implementation AKLog
{
    NSString *_message;
    AKParameter *_parameter;
    NSTimeInterval _timeInterval;
}
- (instancetype)initWithMessage:(NSString *)message
                      parameter:(AKParameter *)parameter
                   timeInterval:(NSTimeInterval)timeInterval
{
    self = [super initWithString:[self operationName]];
    
    if (self) {
        _message = message;
        _parameter = parameter;
        _timeInterval = timeInterval;
        self.state = @"connectable";
        self.dependencies = @[parameter];
    }
    return self;
}

- (NSString *)stringForCSD
{
    return [NSString stringWithFormat:
            @"\nprintks \"%@ %%f\", %f, AKControl(%@)\n",
            _message, _timeInterval, _parameter];
}

- (NSString *)description {
    return @"log";
}
@end

File: AudioKit/Operations/Math/AKLog.swift
//
//  AKLog.swift
//  AudioKit
//
//  Created by Aurelius Prochazka on 9/14/15.
//  Copyright © 2015 AudioKit. All rights reserved.
//

import Foundation

/** Natural log helper function */
func log(parameter: AKParameter) -> AKLog {
    return AKLog(input: parameter)
}

/** Log of the input signal.
*/
@objc class AKLog : AKParameter {
    
    // MARK: - Properties
    
    /** Input to the mathematical function */
    private var input = AKParameter()
    
    // MARK: - Initializers
    
    /** Instantiates the log
    - parameter input: Input signal.
    */
    init(input sourceInput: AKParameter)
    {
        super.init()
        input = sourceInput
        dependencies = [input]
    }
    
    /** Computation of the next value */
    override func compute() {
        leftOutput  = log(input.leftOutput)
        rightOutput = log(input.rightOutput)
    }
    
}
